### PR TITLE
GH-3154: Support `UriBuilderFactory.EncodingMode`

### LINKED ISSUE #3154: URI Is Being Generated With URI Variables Before RestTemplate Does It
https://github.com/spring-projects/spring-integration/blob/cf01b7c0e5007f7da1094230ddcc08b17acac6af/spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java#L301

Spring Integration is generating the entire URI with the declared uri variables, and then passing the generated URI to the RestTemplate exchange method defined in HttpRequestExecutingMessageHandler.exchange()

Here is a truncated declaration of the outbound gateway with the LANGUAGE query parameter templated
```
<int-http:outbound-gateway
                url="/properties/availability?language={LANGUAGE}"                
                rest-template="restTemplate"
        >            
            <int-http:uri-variable name="LANGUAGE" expression="headers['LANGUAGE']"/>            
        </int-http:outbound-gateway>
```

This is causing issues in Micrometer which auto instruments the RestTemplate, and is creating a new Timer for every request as the path/query parameters differ. See https://github.com/micrometer-metrics/micrometer/issues/98 for the issue.

From the above we would expect metrics to be recorded like

`httpClientRequests{clientName="test",method="GET",status="200",uri="/properties/availability&language={LANGUAGE}",quantile="0.5"}`

However what we are seeing is

`httpClientRequests{clientName="test",method="GET",status="200",uri="/properties/availability&language=en-US",quantile="0.5"}
httpClientRequests{clientName="test",method="GET",status="200",uri="/properties/availability&language=de-DE",quantile="0.5"}
httpClientRequests{clientName="test",method="GET",status="200",uri="/properties/availability&language=ko-KR",quantile="0.5"}
........
`

I believe the correct behavior should be that the URI and the URI variables should be passed to the HttpRequestExecutingMessageHandler.exchange() method, so that RestTemplate can map the variables to the URI. This would correctly create the Micrometer timers for the RestTemplate

The call currently is

`httpResponse = this.restTemplate.exchange(uri, httpMethod, httpRequest,
						(ParameterizedTypeReference<?>) expectedResponseType);

`

but it should be 

`httpResponse = this.restTemplate.exchange(uri, httpMethod, httpRequest,
						(ParameterizedTypeReference<?>) expectedResponseType, uriVariables);
`

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-integration/issues/3154

Spring Framework now provides a `DefaultUriBuilderFactory.EncodingMode`
for encoding URIs in the `RestTemplate` before and after uri template
enrichment with uri variables.
Therefore `encodeUri` and manual uri variables substitution is not necessary
in Spring Integration HTTP components

* Deprecate `AbstractHttpRequestExecutingMessageHandler.encodeUri` in favor of
`DefaultUriBuilderFactory.EncodingMode` and respective configuration
on the `RestTemplate` in HTTP module and `WebClient` in WebFlux module

### CODE REVIEW

**Comment by artembilan** [comment]
TODO:

* The same support for WebFlux module
* Docs
* Migration Guide
* Clean up affected test cases for code style and JUnit 5

Will address them lately, but you can review now.

Thanks

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java [hunk 1]
```diff
@@ -118,6 +115,7 @@
 	public AbstractHttpRequestExecutingMessageHandler(Expression uriExpression) {
 		Assert.notNull(uriExpression, "URI Expression is required");
 		this.uriExpression = uriExpression;
+		this.uriFactory.setEncodingMode(DefaultUriBuilderFactory.EncodingMode.TEMPLATE_AND_VALUES);
```
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java]: Why is this needed? This is the default.

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java [hunk 2]
```diff
@@ -85,6 +82,8 @@
 	private static final List<HttpMethod> NO_BODY_HTTP_METHODS =
 			Arrays.asList(HttpMethod.GET, HttpMethod.HEAD, HttpMethod.TRACE);
 
+	DefaultUriBuilderFactory uriFactory = new DefaultUriBuilderFactory();
```
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java]: Why package visibility?

Why not `UriBuilderFactory`?

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java [hunk 3]
```diff
@@ -127,9 +125,25 @@ public AbstractHttpRequestExecutingMessageHandler(Expression uriExpression) {
...
+						: DefaultUriBuilderFactory.EncodingMode.NONE);
+	}
+
+	/**
+	 * Set the encoding mode to use.
+	 * By default this is set to {@link DefaultUriBuilderFactory.EncodingMode#TEMPLATE_AND_VALUES}.
+	 * @param encodingMode the mode to use for uri encoding
+	 * @since 5.3
+	 */
+	public void setEncodingMode(DefaultUriBuilderFactory.EncodingMode encodingMode) {
```
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java]: Why not `setUriBuilderFactory` ?

**Review by garyrussell — CHANGES_REQUESTED** [review]
A few issues/questions.

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java (hunk 2 above)
Comment (artembilan) [reply spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java]: The `UriBuilderFactory` doesn't have a `setEncodingMode()`

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java (hunk 3 above)
Comment (artembilan) [reply spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java]: Right. I'm going to document that for more complex scenarios it would be better to configure an external `RestTemplate` and already use its `setUriTemplateHandler()`.

I would even go with all the `RestTempalte` options as an external reference. I'm not a fun of mutating internal objects.
But that would be too drastic even in the point release...

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java (hunk 2 above)
Comment (garyrussell) [reply spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java]: and package?

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java (hunk 3 above)
Comment (garyrussell) [reply spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java]: Ok, makes sense; proceed with the remaining work 😄

File: spring-integration-http/src/test/java/org/springframework/integration/http/config/HttpOutboundChannelAdapterParserTests.java [hunk 4]
```diff
@@ -192,13 +200,15 @@ public void withUrlAndTemplate() {
...
+		HttpRequestExecutingMessageHandler handler = (HttpRequestExecutingMessageHandler) endpointAccessor
+				.getPropertyValue("handler");
 		DirectFieldAccessor handlerAccessor = new DirectFieldAccessor(handler);
 		assertThat(handlerAccessor.getPropertyValue("expectReply")).isEqualTo(false);
 		assertThat(endpointAccessor.getPropertyValue("inputChannel"))
 				.isEqualTo(this.applicationContext.getBean("requests"));
 		assertThat(handlerAccessor.getPropertyValue("outputChannel")).isNull();
-		DirectFieldAccessor templateAccessor = new DirectFieldAccessor(handlerAccessor.getPropertyValue("restTemplate"));
+		DirectFieldAccessor templateAccessor = new DirectFieldAccessor(handlerAccessor.getPropertyValue("restTemplate"
+		));
```
Comment (garyrussell) [inline spring-integration-http/src/test/java/org/springframework/integration/http/config/HttpOutboundChannelAdapterParserTests.java]: Unusual formatting.

File: spring-integration-http/src/test/java/org/springframework/integration/http/config/HttpOutboundChannelAdapterParserTests.java [hunk 5]
```diff
@@ -224,13 +234,15 @@ public void withUrlExpression() {
...
```
Comment (garyrussell) [inline spring-integration-http/src/test/java/org/springframework/integration/http/config/HttpOutboundChannelAdapterParserTests.java]: Here too.

File: spring-integration-http/src/test/java/org/springframework/integration/http/config/HttpOutboundChannelAdapterParserTests.java [hunk 6]
```diff
@@ -257,13 +269,15 @@ public void withUrlExpressionAndTemplate() {
...
```
Comment (garyrussell) [inline spring-integration-http/src/test/java/org/springframework/integration/http/config/HttpOutboundChannelAdapterParserTests.java]: And here

**Review by garyrussell — CHANGES_REQUESTED** [review]

**Comment by garyrussell** [comment]
> Task :spring-integration-http:checkstyleTest FAILED
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-http/src/test/java/org/springframework/integration/http/config/HttpOutboundChannelAdapterParserTests.java:22:8: Unused import - java.nio.charset.Charset. [UnusedImports]

**Review by garyrussell — APPROVED** [review]
